icu4c/source/i18n/dtfmtsym.cpp

```diff
@@ -86,6 +86,29 @@ static const char16_t gPatternChars[] = {
     0
 };
 
+static const int8_t gLookupPatternChars[] = {
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    //
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    //       !   "   #   $   %   &   '   (   )   *   +   ,   -   .   /
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+#if UDAT_HAS_PATTERN_CHAR_FOR_TIME_SEPARATOR
+    //   0   1   2   3   4   5   6   7   8   9   :   ;   <   =   >   ?
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 37, -1, -1, -1, -1, -1,
+#else
+    //   0   1   2   3   4   5   6   7   8   9   :   ;   <   =   >   ?
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+#endif
+    //   @   A   B   C   D   E   F   G   H   I   J   K   L   M   N   O
+        -1, 22, 36, -1, 10,  9, 11,  0,  5, -1, -1, 16, 26,  2, -1, 31,
+    //   P   Q   R   S   T   U   V   W   X   Y   Z   [   \   ]   ^   _
+        -1, 27, -1,  8, -1, 30, 29, 13, 32, 18, 23, -1, -1, -1, -1, -1,
+    //   `   a   b   c   d   e   f   g   h   i   j   k   l   m   n   o
+        -1, 14, 35, 25,  3, 19, -1, 21, 15, -1, -1,  4, -1,  6, -1, -1,
+    //   p   q   r   s   t   u   v   w   x   y   z   {   |   }   ~
+        -1, 28, 34,  7, -1, 20, 24, 12, 33,  1, 17, -1, -1, -1, -1, -1
+};
+
```

Comment:
I'm worried, again, about the possibility of somebody in the future modifying `gPatternChars` and not knowing they also have to modify this table.  Again, that doesn't seem terribly likely (probably not likely at all), but it worries me a little.

My first response was to ask if you could generate this table on the fly the first time it's used for something, but the more I think about it, the more I think that's probably overkill.  I'd be happy with a comment above this declaration explaining what it is and how it was derived.

Reply:
This would be much cleaner with designated initializers but C++ does not support them (the compiler accepts them but I believe `-Werror` is enabled so it fails in the CI). I've found the pattern characters are mentioned in many tables like this one (I got this template from `SimpleDateFormat::getLevelFromChar`). Alternatively, `constexpr` could be used. I'll settle for the comment for now, and just out of curiosity I'll see how feasible it would be to generate it.

Reply:
A simple comment is fine with me.  The more I think about it, the more I think generating the table on the fly isn't worth the trouble, considering the remote likelihood of its contents changing.